cluster/service.go

```
@@ -58,7 +59,31 @@ func New(cfg Config, svrMetrics *monitoring.GRPCServerMetrics) *Service {
 
 	fsm := NewFSM(cfg, prometheus.DefaultRegisterer)
 	raft := NewRaft(cfg.NodeSelector, &fsm, client)
-	replicationEngine := replication.NewShardReplicationEngine(cfg.Logger, cfg.NodeSelector.LocalName(), fsm.replicationManager.GetReplicationFSM(), raft, cfg.ReplicaCopier)
+	fsmOpProducer := replication.NewFSMOpProducer(
+		cfg.Logger.WithFields(logrus.Fields{"component": "replication_engine_producer", "node": cfg.NodeSelector.LocalName()}),
+		fsm.replicationManager.GetReplicationFSM(),
+		5*time.Second,
+		cfg.NodeSelector.LocalName(),
+	)
+	retentionPolicy := replication.NewTTLOpRetentionPolicy(
+		5*24*time.Hour,
+		replication.RealTimer{},
+	)
+	realTimeProvider := replication.RealTimeProvider{}
+	tracker := replication.NewOpTracker(realTimeProvider)
+	replicaCopyOpConsumer := replication.NewCopyOpConsumer(
+		cfg.Logger.WithField("component", "replication_engine_consumer").WithField("node", cfg.NodeSelector.LocalName()),
+		raft,
+		cfg.ReplicaCopier,
+		realTimeProvider,
+		cfg.NodeSelector.LocalName(),
+		tracker,
+		backoff.NewConstantBackOff(5*time.Second),
```

Two question here:
1) do we need to add some maximum number of retries here to avoid retrying forever?
2) for now, should we just abort the op if there is an error instead of retrying it?

---

I honestly did it the first time...including 3 retries as the default. Then my idea was...we need to distinguish between permanent failures and temporary failures to avoid retrying in case there are permanent failures. I think retrying if there are permanent failures is pointless and a waste of resources. Since we do not have this ability yes (distinguish permanent versus temporary failures) I decided to leave out the retry logic just to avoid more complexity and replication operations being retried without control. In my opinion if we decide to retry operations we need to make that distinction....otherwise better to fail soon rather than later. For now, I would lean more towards the idea of "fail fast"...at least until we know that this is working reliably enough to do something more sophisticated. Then maybe we can add a retry count in the api request too, perhaps?

---

ok, can you create a todo/task to make sure we think more about this before release?